File: packages/aps-cli-node/src/parsers/adaptor.ts
import fs from 'node:fs/promises';

/**
 * Parsed constant value from adaptor.md.
 */
export type ConstantValue = string | number | boolean | unknown[] | Record<string, unknown>;

/**
 * Parsed format contract from adaptor.md.
 */
export interface FormatContract {
  id: string;
  name: string;
  purpose: string;
  body: string;
}

/**
 * Complete parsed adaptor.md data.
 */
export interface AdaptorData {
  instructions: string;
  constants: Record<string, ConstantValue>;
  formats: Record<string, FormatContract>;
}

const SECTION_RE = /<(instructions|constants|formats)>([\s\S]*?)<\/\1>/g;
const FORMAT_BLOCK_RE = /<format\b([^>]*)>([\s\S]*?)<\/format>/g;
const FORMAT_ATTR_RE = /\b(id|name|purpose)="([^"]*)"/g;
const NUMBER_RE = /^-?(?:\d+|\d*\.\d+)(?:[eE][+-]?\d+)?$/;

function splitLines(raw: string): string[] {
  return raw.split(/\r\n|\n|\r/);
}

function parseIdentifierArray(raw: string): string[] | null {
  const s = raw.trim();
  if (!s.startsWith('[') || !s.endsWith(']')) return null;

  const inner = s.slice(1, -1).trim();
  if (!inner) return [];

  const parts = inner
    .split(',')
    .map((p) => p.trim())
    .filter(Boolean);

  const out: string[] = [];
  for (const part of parts) {
    const quoted =
      (part.startsWith('"') && part.endsWith('"')) ||
      (part.startsWith("'") && part.endsWith("'"));
    if (quoted) {
      out.push(part.slice(1, -1));
      continue;
    }

    if (!/^[A-Z][A-Z0-9_]*$/.test(part)) return null;
    out.push(part);
  }

  return out;
}

/**
 * Parse a CSV string into an array of objects keyed by header fields.
 */
function parseCsvBlock(csv: string): Record<string, string>[] {
  const lines = splitLines(csv.trim()).filter((l) => l.trim());
  if (lines.length < 1) return [];

  const headers = splitCsvRow(lines[0] ?? '');
  const rows: Record<string, string>[] = [];

  for (let i = 1; i < lines.length; i++) {
    const cells = splitCsvRow(lines[i] ?? '');
    const row: Record<string, string> = {};
    for (let j = 0; j < headers.length; j++) {
      row[headers[j] ?? ''] = cells[j] ?? '';
    }
    rows.push(row);
  }

  return rows;
}

/**
 * Split a CSV row respecting quoted fields.
 */
function splitCsvRow(line: string): string[] {
  const cells: string[] = [];
  let current = '';
  let inQuote = false;

  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (inQuote) {
      if (ch === '"' && line[i + 1] === '"') {
        current += '"';
        i++;
      } else if (ch === '"') {
        inQuote = false;
      } else {
        current += ch;
      }
    } else {
      if (ch === '"') {
        inQuote = true;
      } else if (ch === ',') {
        cells.push(current);
        current = '';
      } else {
        current += ch;
      }
    }
  }

  cells.push(current);
  return cells;
}

/**
 * Parse the constants section of an adaptor.md file.
 */
function parseConstants(raw: string): Record<string, ConstantValue> {
  const constants: Record<string, ConstantValue> = {};
  const lines = splitLines(raw);
  let i = 0;

  while (i < lines.length) {
    const line = lines[i] ?? '';
    const trimmed = line.trim();

    if (!trimmed || trimmed.startsWith('//') || trimmed.startsWith('#')) {
      i++;
      continue;
    }

    const colonIdx = trimmed.indexOf(':');
    if (colonIdx === -1) {
      i++;
      continue;
    }

    const key = trimmed.slice(0, colonIdx).trim();
    const rest = trimmed.slice(colonIdx + 1).trim();

    const blockMatch = rest.match(/^(JSON|TEXT|CSV|YAML)<<$/);
    if (blockMatch) {
      const blockType = blockMatch[1];
      const bodyLines: string[] = [];
      i++;
      while (i < lines.length) {
        const bLine = lines[i] ?? '';
        if (bLine.trim() === '>>') {
          i++;
          break;
        }
        bodyLines.push(bLine);
        i++;
      }

      const body = bodyLines.join('\n');

      if (blockType === 'JSON') {
        try {
          constants[key] = JSON.parse(body) as ConstantValue;
        } catch {
          constants[key] = body;
        }
      } else if (blockType === 'CSV') {
        constants[key] = parseCsvBlock(body);
      } else {
        constants[key] = body;
      }

      continue;
    }

    if (rest.startsWith('[')) {
      try {
        constants[key] = JSON.parse(rest) as ConstantValue;
      } catch {
        const parsed = parseIdentifierArray(rest);
        constants[key] = parsed ?? rest;
      }
      i++;
      continue;
    }

    if (
      (rest.startsWith('"') && rest.endsWith('"')) ||
      (rest.startsWith("'") && rest.endsWith("'"))
    ) {
      constants[key] = rest.slice(1, -1);
      i++;
      continue;
    }

    if (rest === 'true') {
      constants[key] = true;
      i++;
      continue;
    }
    if (rest === 'false') {
      constants[key] = false;
      i++;
      continue;
    }

    if (NUMBER_RE.test(rest)) {
      const num = Number(rest);
      if (!Number.isNaN(num)) {
        constants[key] = num;
        i++;
        continue;
      }
    }

    constants[key] = rest;
    i++;
  }

  return constants;
}

/**
 * Parse the formats section of an adaptor.md file.
 */
function parseFormats(raw: string): Record<string, FormatContract> {
  const formats: Record<string, FormatContract> = {};
  let match: RegExpExecArray | null;
  FORMAT_BLOCK_RE.lastIndex = 0;

  while ((match = FORMAT_BLOCK_RE.exec(raw)) !== null) {
    const attrsRaw = match[1] ?? '';
    const body = (match[2] ?? '').trim();

    const attrs: Record<string, string> = {};
    let a: RegExpExecArray | null;
    FORMAT_ATTR_RE.lastIndex = 0;
    while ((a = FORMAT_ATTR_RE.exec(attrsRaw)) !== null) {
      const k = a[1];
      if (!k) continue;
      attrs[k] = a[2] ?? '';
    }

    const id = attrs['id'];
    if (!id) continue;

    formats[id] = {
      id,
      name: attrs['name'] ?? '',
      purpose: attrs['purpose'] ?? '',
      body,
    };
  }

  return formats;
}

/**
 * Parse an adaptor.md file into structured data.
 */
export async function parseAdaptorMd(filePath: string): Promise<AdaptorData> {
  const raw = await fs.readFile(filePath, 'utf8');
  return parseAdaptorMdString(raw);
}

/**
 * Parse an adaptor.md string into structured data.
 */
export function parseAdaptorMdString(raw: string): AdaptorData {
  const data: AdaptorData = {
    instructions: '',
    constants: {},
    formats: {},
  };

  let match: RegExpExecArray | null;
  SECTION_RE.lastIndex = 0;

  while ((match = SECTION_RE.exec(raw)) !== null) {
    const section = match[1];
    const content = match[2] ?? '';

    switch (section) {
      case 'instructions':
        data.instructions = content.trim();
        break;
      case 'constants':
        data.constants = parseConstants(content);
        break;
      case 'formats':
        data.formats = parseFormats(content);
        break;
    }
  }

  return data;
}

/**
 * Extract a string constant or return a default.
 */
export function getString(
  constants: Record<string, ConstantValue>,
  key: string,
  fallback = ''
): string {
  const v = constants[key];
  return typeof v === 'string' ? v : fallback;
}

/**
 * Extract a string array constant or return empty array.
 */
export function getStringArray(constants: Record<string, ConstantValue>, key: string): string[] {
  const v = constants[key];
  if (Array.isArray(v)) {
    return v.filter((item): item is string => typeof item === 'string');
  }
  return [];
}

File: packages/aps-cli-py/src/aps_cli/parsers/adaptor.py
"""Parse adaptor.md files into structured data.

Handles <instructions>, <constants>, and <formats> sections including
TEXT, JSON, CSV, and YAML block constant types.
"""

from __future__ import annotations

import csv
import io
import json
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Union

ConstantValue = Union[str, int, float, bool, list, dict]

SECTION_RE = re.compile(r"<(instructions|constants|formats)>(.*?)</\1>", re.DOTALL)
FORMAT_BLOCK_RE = re.compile(r"<format\b([^>]*)>(.*?)</format>", re.DOTALL)
FORMAT_ATTR_RE = re.compile(r'\b(id|name|purpose)="([^"]*)"')
NUMBER_RE = re.compile(r"^-?(?:\d+|\d*\.\d+)(?:[eE][+-]?\d+)?$")
INTEGER_RE = re.compile(r"^-?\d+$")


@dataclass(frozen=True)
class FormatContract:
    """A parsed format block from adaptor.md."""

    id: str
    name: str
    purpose: str
    body: str


@dataclass
class AdaptorData:
    """Complete parsed adaptor.md content."""

    instructions: str = ""
    constants: dict[str, ConstantValue] = field(default_factory=dict)
    formats: dict[str, FormatContract] = field(default_factory=dict)


def _split_csv_row(line: str) -> list[str]:
    """Split a CSV row respecting quoted fields."""
    reader = csv.reader(io.StringIO(line))
    for row in reader:
        return list(row)
    return []


def _parse_csv_block(body: str) -> list[dict[str, str]]:
    """Parse CSV block into list of dicts keyed by header."""
    lines = [l for l in body.strip().splitlines() if l.strip()]
    if not lines:
        return []

    headers = _split_csv_row(lines[0])
    rows: list[dict[str, str]] = []
    for line in lines[1:]:
        cells = _split_csv_row(line)
        row = {headers[j]: (cells[j] if j < len(cells) else "") for j in range(len(headers))}
        rows.append(row)
    return rows


def _parse_identifier_array(raw: str) -> list[str] | None:
    s = raw.strip()
    if not (s.startswith("[") and s.endswith("]")):
        return None

    inner = s[1:-1].strip()
    if not inner:
        return []

    parts = [p.strip() for p in inner.split(",") if p.strip()]
    out: list[str] = []
    for part in parts:
        if (part.startswith('"') and part.endswith('"')) or (part.startswith("'") and part.endswith("'")):
            out.append(part[1:-1])
            continue

        if not re.match(r"^[A-Z][A-Z0-9_]*$", part):
            return None
        out.append(part)

    return out


def _parse_constants(raw: str) -> dict[str, ConstantValue]:
    """Parse the constants section text."""
    constants: dict[str, ConstantValue] = {}
    lines = raw.splitlines()
    i = 0

    while i < len(lines):
        line = lines[i]
        trimmed = line.strip()

        if not trimmed or trimmed.startswith("//") or trimmed.startswith("#"):
            i += 1
            continue

        colon_idx = trimmed.find(":")
        if colon_idx == -1:
            i += 1
            continue

        key = trimmed[:colon_idx].strip()
        rest = trimmed[colon_idx + 1 :].strip()

        block_match = re.match(r"^(JSON|TEXT|CSV|YAML)<<$", rest)
        if block_match:
            block_type = block_match.group(1)
            body_lines: list[str] = []
            i += 1
            while i < len(lines):
                if lines[i].strip() == ">>":
                    i += 1
                    break
                body_lines.append(lines[i])
                i += 1
            body = "\n".join(body_lines)

            if block_type == "JSON":
                try:
                    constants[key] = json.loads(body)
                except json.JSONDecodeError:
                    constants[key] = body
            elif block_type == "CSV":
                constants[key] = _parse_csv_block(body)
            else:
                constants[key] = body
            continue

        if rest.startswith("["):
            try:
                constants[key] = json.loads(rest)
            except json.JSONDecodeError:
                parsed = _parse_identifier_array(rest)
                constants[key] = parsed if parsed is not None else rest
            i += 1
            continue

        if (rest.startswith('"') and rest.endswith('"')) or (rest.startswith("'") and rest.endswith("'")):
            constants[key] = rest[1:-1]
            i += 1
            continue

        if rest == "true":
            constants[key] = True
            i += 1
            continue
        if rest == "false":
            constants[key] = False
            i += 1
            continue

        if NUMBER_RE.match(rest):
            try:
                num = float(rest)
                if INTEGER_RE.match(rest):
                    constants[key] = int(num)
                else:
                    constants[key] = num
                i += 1
                continue
            except ValueError:
                pass

        constants[key] = rest
        i += 1

    return constants


def _parse_formats(raw: str) -> dict[str, FormatContract]:
    """Parse the formats section text."""
    formats: dict[str, FormatContract] = {}

    for m in FORMAT_BLOCK_RE.finditer(raw):
        attrs_raw = m.group(1) or ""
        body = (m.group(2) or "").strip()

        attrs: dict[str, str] = {}
        for a in FORMAT_ATTR_RE.finditer(attrs_raw):
            attrs[a.group(1)] = a.group(2) or ""

        fid = attrs.get("id")
        if not fid:
            continue

        formats[fid] = FormatContract(
            id=fid,
            name=attrs.get("name", ""),
            purpose=attrs.get("purpose", ""),
            body=body,
        )

    return formats


def parse_adaptor_md(file_path: Path) -> AdaptorData:
    """Parse an adaptor.md file into structured data."""
    raw = file_path.read_text(encoding="utf-8")
    return parse_adaptor_md_string(raw)


def parse_adaptor_md_string(raw: str) -> AdaptorData:
    """Parse adaptor.md content string into structured data."""
    data = AdaptorData()

    for m in SECTION_RE.finditer(raw):
        section = m.group(1)
        content = m.group(2) or ""

        if section == "instructions":
            data.instructions = content.strip()
        elif section == "constants":
            data.constants = _parse_constants(content)
        elif section == "formats":
            data.formats = _parse_formats(content)

    return data


def get_string(constants: dict[str, Any], key: str, fallback: str = "") -> str:
    """Extract a string constant or return a default."""
    v = constants.get(key)
    return v if isinstance(v, str) else fallback


def get_string_array(constants: dict[str, Any], key: str) -> list[str]:
    """Extract a string array constant or return empty list."""
    v = constants.get(key)
    if isinstance(v, list):
        return [item for item in v if isinstance(item, str)]
    return []